### PR TITLE
feat: Add NPU support for fsdp2 and fix compatibility issues

### LINKED ISSUE #337: [RFC] Add support for megatron backend in Ascend NPU
This issue proposes adding **optional Ascend NPU support** to ROLL by integrating **[MindSpeed](https://gitcode.com/Ascend/MindSpeed)**, which is essentially a toolkit to run megatron on ascend npu.

#### Motivation

On Ascend NPU, MindSpeed provides Megatron-compatible patches and runtime extensions.
ROLL currently does not expose a clean way to apply these patches or pass the required configuration, which makes Ascend execution difficult.

#### High-level approach

* Apply MindSpeed patches **before Megatron modules are initialized**
* Expose a small set of **MindSpeed-required configuration fields** at the config layer
* Allow MindSpeed-related arguments to pass through existing ROLL configs
* Keep all behavior **opt-in and limited to Ascend NPU**

#### Non-goals

* No behavior changes for CUDA / CPU
* No MindSpeed hard dependency
* No architectural changes to Megatron

#### Environment

* Target: Ascend NPU
* MindSpeed version: 0.14.0

### PR DESCRIPTION
**Pull request by UsernameFull**
## What does this PR do?

> Fixes #337 

### 1. Adapt FSDP2 for NPU backend
1. Ensure parity: Invoke `torch.npu.empty_cache()` when running on NPU devices.
2. Remove unused dependency on `flash_attn.bert_padding`
3. Move CPU weights to `current_platform.device_type` before broadcasting

### 2. Additional NPU compatibility features
1. Fix NPU crash by replacing `_set_allocator_settings()` with `PYTORCH_NPU_ALLOC_CONF`.

### 3. Bug fix
1. **Fix validation batch size calculation**: Removed the ga_steps multiplier from global_val_batch_size. Gradient accumulation steps should not inflate the validation batch size—gradient accumulation is not performed during validation.
2. Replace `get_data_modulo_expert_parallel_rank` with `get_expert_data_parallel_rank` as the former is deprecated.
3. Change the initialization of `sequence_packing_args` from `default=SequencePackingConfig()` to `default_factory=SequencePackingConfig`.

## result

Critic/score/mean curves for the same config running on GPU vs. NPU

```yaml
defaults:
  - ../../examples/config/envs@_here_
  - ../../examples/config/deepspeed_zero@_here_
  - ../../examples/config/deepspeed_zero2@_here_
  - ../../examples/config/deepspeed_zero3@_here_
  - ../../examples/config/deepspeed_zero3_cpuoffload@_here_


hydra:
  run:
    dir: .
  output_subdir: null

pg_variant: ppo # topr, vanilla, tis, cispo, kimi15, ppo
exp_name: Qwen3-8B-RLVR-${pg_variant}
seed: 42
logging_dir: ./output/logs
output_dir: ./output
system_envs:
  USE_MODELSCOPE: '1'

checkpoint_config:
  type: file_system
  output_dir: ./ckpt


num_gpus_per_node: 8

max_steps: 100
save_steps: 100
logging_steps: 1
eval_steps: 10
resume_from_checkpoint: false


rollout_batch_size: 64  # prompt
prompt_length: 1024
response_length: 2048

num_return_sequences_in_group: 8
ppo_epochs: 1
adv_estimator: "reinforce"

# clip
value_clip: 0.5
reward_clip: 10
advantage_clip: 2.0
dual_clip_loss: true

# normalize
norm_mean_type: batch
norm_std_type: batch

# data mask
max_len_mask: true
difficulty_mask: true
difficulty_low_threshold: 0.2
difficulty_high_threshold: 0.95
error_max_len_clip: false

# data weight
difficulty_loss_weight: false
length_loss_weight: false

# reward
add_token_level_kl: false

# advantage
whiten_advantages: true


pretrain: Qwen/Qwen3-8B
reward_pretrain: Qwen/Qwen3-8B

validation:
  data_args:
    template: qwen3
    file_name:
      - data/math_benchmarks.jsonl
  generating_args:
    top_p: 0.6
    top_k: 50
    num_beams: 1
    temperature: 0.6
    num_return_sequences: 1
  eval_steps: 10

actor_train:
  worker_cls: roll.pipeline.rlvr.actor_pg_worker.ActorPGWorker
  pg_variant: ppo topr, vanilla, tis, cispo, kimi15, ppo
  model_args:
    flash_attn: fa2
    disable_gradient_checkpointing: false
    dtype: bf16
    model_type: ~
  training_args:
    learning_rate: 1.0e-6
    weight_decay: 0
    per_device_train_batch_size: 1
    gradient_accumulation_steps: 64
    warmup_steps: 20
    num_train_epochs: 50
  data_args:
    template: qwen3
    file_name:
      - data/math_deepmath_deal.jsonl
      - data/llm_judge_Multi-subject-RLVR_deal_new.jsonl
    domain_interleave_probs:
      math_rule: 1
    dataset_dir: data
    messages: messages
    interleave_probs: "1.0"
    preprocessing_num_workers: 16
  strategy_args:
    strategy_name: deepspeed_train
    strategy_config: ${deepspeed_zero3}
  device_mapping: list(range(0,4))
  infer_batch_size: 4

actor_infer:
  model_args:
    flash_attn: fa2
    disable_gradient_checkpointing: true
    dtype: bf16
  generating_args:
    max_new_tokens: ${response_length}
    top_p: 0.99
    top_k: 100
    num_beams: 1
    temperature: 0.99
    num_return_sequences: ${num_return_sequences_in_group}
  data_args:
    template: qwen3
  strategy_args:
     strategy_name: vllm
     strategy_config:
       gpu_memory_utilization: 0.6
       block_size: 16
       max_model_len: 8000
  device_mapping: list(range(4,6))
  infer_batch_size: 1

reference:
  model_args:
    flash_attn: fa2
    disable_gradient_checkpointing: true
    dtype: bf16
    model_type: ~
  data_args:
    template: qwen3
  strategy_args:
    strategy_name: hf_infer
    strategy_config: ~
  device_mapping: list(range(6,8))
  infer_batch_size: 8

rewards:
  math_rule:
    worker_cls: roll.pipeline.rlvr.rewards.math_rule_reward_worker.MathRuleRewardWorker
    model_args:
      model_name_or_path: ${reward_pretrain}
    data_args:
      template: qwen3
    tag_included: [deepmath_103k, 'MATH-500', 'OlympiadBench', 'minervamath', 'aime2025', 'gsm8k', 'aime', 'amc23', 'math_rule']
    world_size: 8
    infer_batch_size: 1
```

image

### CODE REVIEW

File: roll/distributed/strategy/fsdp2_strategy.py [hunk 1]
```diff
@@ -1249,7 +1259,10 @@ def train_step(
                     self.scheduler.step()
                     self.optimizer.zero_grad(set_to_none=True)
 
-        torch.cuda.empty_cache()
+        if current_platform.device_type == "cuda":
```
[the remark this answers is not in the document]
Comment (UsernameFull) [reply roll/distributed/strategy/fsdp2_strategy.py]: Changed to `current_platform.empty_cache` as suggested.
